src/jit/lower.cpp

```diff
@@ -2427,12 +2430,27 @@ void Lowering::LowerCompare(GenTree* cmp)
                 // situations. In particular this include CALL, sometimes the JIT unnecessarilly widens
                 // the result of bool returning calls.
                 //
+                bool removeCast =
+#ifdef _TARGET_ARM64_
+                    (op2Value <= ((castToType == TYP_BOOL) ? 1 : 0)) && cmp->OperIs(GT_EQ, GT_NE, GT_GT) &&
```

Comment:
This assumes that bools are always 0 or 1, this isn't actually guaranteed. However, the JIT already makes this assumption in a few other places so this is probably fine.

Reply:
I was actually surprised the cast to BOOL does not force this given the performance warnings I get for similar C++ code.

Reply:
In C++/C, cast to `bool(a) === (a == 0) ? false : true`.  As a C/C++ programmer I would expect C# to behave the same.

https://gcc.godbolt.org/

These result in identical code
```
bool Test(int b) 
{ 
    return b; 
}

bool TestCast(int b) 
{ 
    return bool(b); 
}
```
Looking at Test()
Arm64 gcc 7.2
```
Test(int):
  sub sp, sp, #16
  str w0, [sp, 12]
  ldr w0, [sp, 12]
  cmp w0, 0
  cset w0, ne
  uxtb w0, w0
  add sp, sp, 16
  ret
```

x86-64 gcc 7.2
```
Test(int):
  push rbp
  mov rbp, rsp
  mov DWORD PTR [rbp-4], edi
  cmp DWORD PTR [rbp-4], 0
  setne al
  pop rbp
  ret
```

Reply:
> In C++/C, cast to bool(a) === (a == 0) ? false : true. As a C/C++ programmer I would expect C# to behave the same.

It's complicated. IL doesn't really have the notion of `bool` so there's no such thing as cast to `bool`. There are only the `i1` and `u1` byte sized integer types and the associated casts.

`bool` does exist in metadata so a method returning `bool` is not the same thing as a method returning `u1`/`i1`. Same for fields.

The JIT has this special `TYP_BOOL` type which is mostly used as a hint that a metadata `bool` was encountered. This information is occasionally used to infer that the value can be only 0/1 but in fact this is not guaranteed. It is difficult to produce non-0/1 bools in C# but not impossible (unsafe code, unions etc.). See https://github.com/dotnet/coreclr/issues/12983#issuecomment-317201639 for some additional details regarding `TYP_BOOL` uses.

In the JIT a cast to `TYP_BOOL` is really a cast to `TYP_UBYTE`. That is, zero extend the byte to int32.

Reply:
That x64 code produced by gcc, only sets the low 8 bits:

'setne al'

The upper 24-bits are left as  random junk.  For Arm64 the C compiler smartly zero extends to  the full width of the register using 

'uxtb w0, w0'

Reply:
> For Arm64 the C compiler smartly zero extends to the full width of the register using 
`uxtb w0, w0`

Actually the `uxtb w0, w0` is redundant.

After
```
  cmp w0, 0
  cset w0, ne
```

x0 and w0 is guaranteed to contain either 0 or 1.

Reply:
OK, but in the JIT due to that x64 C compiler behavior we have to zero extend the return value before using it after every call that returns a byte/bool.

Reply:
There are 2 separate issues here:
* whether the widening cast is required or not
* whether `TYP_BOOL` implies that 0 and 1 are the only possible values

My initial comment was about the later, it is not guaranteed that a bool method returns only 0 and 1. It may return 42:
```
ldc.i4 42
ret
```

Reply:
> we have to zero extend the return value before using it after every call that returns a byte/bool.

Understood.  @mikedn optimization which is in tip removes the cast and converts it to a byte compare.  This is doing the same for arm64.  Perhaps we also should not add the cast in importer for arm64?

Reply:
Of course `TYP_BOOL(42) == true` is `false` in C#, but true in C++ 

`TYP_BOOL(43) == true` is `false` in C#, but would be true with this patch as written.

Whether this is a problem, is the crux of the issue.

Reply:
Generally we don't want to do any machine specific pre-optimizations in the importer.

Reply:
> Of course TYP_BOOL(42) == true is false in C#, but true in C++

The language is irrelevant here. In the JIT IR `TYP_BOOL(42) == 42`.

> Of course 42 and 43 are not bools and shouldn't be cast to bool in C#

According to the ECMA spec they are:

> III.1.1.2 Boolean data type A CLI Boolean type occupies 1 byte in memory. A bit pattern of all zeroes denotes a value of false. A bit pattern with any one or more bits set (analogous to a non-zero integer) denotes a value of true. For the purpose of stack operations boolean values are treated as unsigned 1-byte integers (§III.1.1.1).

However, the JIT (and the C# compiler) tend to assume that bools are only 0 and 1. In this particular case this assumption is probably fine. There certainly are situations where assuming 0/1 would be dangerous (e.g. array range check elimination).